contracts/mirror/src/contract.rs

```diff
@@ -0,0 +1,473 @@
+                    }));
+                }
+            }
+        }
+    }
+    Ok(response("complete", CONTRACT_NAME, attrs).add_messages(msgs))
+}
+
+#[cfg_attr(not(feature = "library"), cosmwasm_std::entry_point)]
+pub fn reply(
+    deps: DepsMut<NeutronQuery>,
+    _env: Env,
+    msg: Reply,
+) -> ContractResult<Response<NeutronMsg>> {
+    finalize_bond(deps, msg)
+}
+
+pub fn finalize_bond(
+    deps: DepsMut<NeutronQuery>,
+    msg: Reply,
+) -> ContractResult<Response<NeutronMsg>> {
+    match msg.result {
+        cosmwasm_std::SubMsgResult::Ok(res) => {
+            let mut bond = BONDS.load(deps.storage, msg.id)?;
+            bond.state = BondState::Bonded;
+            // get token factory mint event
+            let tf_mint_event = res
+                .events
+                .iter()
+                .find(|x| x.ty == "tf_mint")
+                .ok_or(ContractError::NoTokensMinted)?;
+            // get amount from mint event
+            let coin = Coin::from_str(
+                &tf_mint_event
+                    .attributes
+                    .iter()
+                    .find(|x| x.key == "amount")
+                    .ok_or(ContractError::NoTokensMintedAmountFound)?
+                    .value,
+            )?;
+            bond.received = Some(coin);
+            BONDS.save(deps.storage, msg.id, &bond)?;
+            let attrs = vec![
+                attr("action", "finalize_bond"),
+                attr("id", msg.id.to_string()),
+                attr("state", bond.state.to_string()),
+            ];
+            Ok(response("finalize_bond", CONTRACT_NAME, attrs))
+        }
+        cosmwasm_std::SubMsgResult::Err(_) => unreachable!(), // as there is only SubMsg::reply_on_success()
+    }
+}
+
+#[cfg_attr(not(feature = "library"), cosmwasm_std::entry_point)]
+pub fn sudo(
+    deps: DepsMut<NeutronQuery>,
+    _env: Env,
+    msg: TransferSudoMsg,
+) -> ContractResult<Response<NeutronMsg>> {
+    match msg {
+        TransferSudoMsg::Response { request, data } => sudo_response(deps, request, data),
+        TransferSudoMsg::Error { request, details } => sudo_error(deps, request, details),
+        TransferSudoMsg::Timeout { request } => sudo_timeout(deps, request),
+    }
+}
+
+fn sudo_error(
+    deps: DepsMut<NeutronQuery>,
+    req: RequestPacket,
+    _data: String,
+) -> ContractResult<Response<NeutronMsg>> {
+    let id = get_id_from_request_memo(&req)?;
+    deps.api.debug(
+        format!(
+            "WASMDEBUG: sudo_error: ack received: {:?} id: {:?}",
+            req, id
+        )
+        .as_str(),
+    );
+    let mut bond = BONDS.load(deps.storage, id)?;
+    bond.state = BondState::Bonded;
+    BONDS.save(deps.storage, id, &bond)?;
+    let attrs = vec![
+        attr("action", "sudo_error"),
+        attr("id", id.to_string()),
+        attr("state", bond.state.to_string()),
+    ];
+    Ok(response("sudo_error", CONTRACT_NAME, attrs))
+}
+
+fn sudo_timeout(
```

Comment:
`sudo_timeout` and `sudo_error` looks similar, maybe we can use more generic function for both of them?

Reply:
not sure. do you mean combine logic into one function?

Reply:
yes